neonvm/runner/main.go

```diff
@@ -532,10 +544,198 @@ func main() {
+	}
+
+	log.Println("cleaning up cgroup...")
+	err = os.Remove(cgroupPath)
+	if err != nil {
+		log.Println(err)
+	}
+	log.Println("exiting")
+}
+
+func handleCPUChange(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		log.Printf("unexpected method: %s\n", r.Method)
+		w.WriteHeader(500)
+		return
+	}
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("could not read body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	parsed := neonApi.VCPUChange{}
+	err = json.Unmarshal(body, &parsed)
+	if err != nil {
+		log.Printf("could not parse body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	cgroupPath := r.Context().Value(cgroupContextKey{})
+	val, ok := cgroupPath.(string)
+	if !ok {
+		log.Println("no path in context")
+		w.WriteHeader(500)
+		return
+	}
+
+	// update cgroup
+	fraction := getCPUFraction(parsed.VCPUs)
+	log.Printf("got CPU update %v (%v)\n", parsed.VCPUs, fraction)
+	err = setCgroupLimit(fraction, val)
+	if err != nil {
+		log.Printf("could not read body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	w.WriteHeader(200)
+}
+
+type cgroupContextKey struct{}
+
+func listenForCPUChanges(port int32, cgroupPath string) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/cpu_change", handleCPUChange)
+	baseCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	baseCtx = context.WithValue(baseCtx, cgroupContextKey{}, cgroupPath)
+	server := http.Server{
+		Addr:              fmt.Sprintf("0.0.0.0:%d", port),
+		Handler:           mux,
+		ReadTimeout:       5 * time.Second,
+		ReadHeaderTimeout: 5 * time.Second,
+		WriteTimeout:      5 * time.Second,
+		BaseContext: func(_ net.Listener) context.Context {
+			return baseCtx
+		},
+	}
+	err := server.ListenAndServe()
+	if errors.Is(err, http.ErrServerClosed) {
+		log.Println("server closed")
+	} else if err != nil {
+		log.Fatalf("error starting server: %s\n", err)
+	}
+}
+
+// isSystemCgroup checks if we have systemd-managed-like cgroup layout
+func isSystemdCgroup() (bool, error) {
+	_, err := os.Stat("/sys/fs/cgroup/systemd")
+	if errors.Is(err, fs.ErrNotExist) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func setCgroupLimit(fraction float64, cgroupPath string) error {
+	isV2, err := isCgroupv2(cgroupPath)
+	if err != nil {
+		return err
+	}
+	if isV2 {
+		data := fmt.Sprintf("%d %d", int(fraction*float64(cgroupPeriod)), cgroupPeriod)
+		err := os.WriteFile(path.Join(cgroupPath, "cpu.max"), []byte(data), 0644)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	err = os.WriteFile(path.Join(cgroupPath, "cpu.cfs_period_us"), []byte(fmt.Sprintf("%d", cgroupPeriod)), 0644)
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(path.Join(cgroupPath, "cpu.cfs_quota_us"), []byte(fmt.Sprintf("%d", int(fraction*float64(cgroupPeriod)))), 0644)
+}
+
+func createCgroup(cgroupPath string) error {
+	err := os.Mkdir(cgroupPath, os.ModeDir)
+	if os.IsExist(err) {
```

Comment:
I'm confused by the decision/motivation for your choices for `os.IsExist` vs `os.IsNotExist` and `errors.Is(err, fs.IsExist)`